File: sdks/jvm/src/main/kotlin/com/heossi/qnsi/StorageClient.kt
package com.heossi.qnsi

import com.heossi.qnsi.internal.Internal
import com.heossi.qnsi.internal.RequestOptions
import com.heossi.qnsi.internal.decodeB64
import com.heossi.qnsi.internal.encodeB64
import com.heossi.qnsi.internal.stringField
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

private const val PREFIX = "/proxy/storage/v1"

/** Object payload + optional content-type / SSE algorithm / metadata for [StorageClient.putObject]. */
public data class PutObjectInput @JvmOverloads constructor(
    public val data: ByteArray,
    public val contentType: String? = null,
    public val sseAlgorithm: String? = null,
    public val metadata: JsonObject? = null,
)

/** Decrypted object payload plus the descriptor JSON returned by the service. */
public data class StorageObject(
    public val data: ByteArray,
    public val descriptor: JsonObject,
)

/** PQC-encrypted object storage with SSE-X. Wraps storage-service `/storage/storage/v1`. */
public class StorageClient internal constructor(private val transport: Internal) {
    @JvmOverloads
    public fun putObject(
        bucket: String,
        key: String,
        input: PutObjectInput,
        idempotencyKey: String? = null,
    ): JsonObject {
        val body = buildJsonObject {
            put("dataB64", encodeB64(input.data))
            input.contentType?.let { put("contentType", it) }
            input.sseAlgorithm?.let { put("sseAlgorithm", it) }
            input.metadata?.let { put("metadata", it) }
        }.toString()
        return transport.requestJson(
            "PUT", "$PREFIX/buckets/$bucket/objects/$key", body,
            RequestOptions(idempotencyKey = idempotencyKey),
        )
    }

    /** Fetch and decrypt an object. Throws [QnsiApiException] if the descriptor omits `dataB64`. */
    public fun getObject(bucket: String, key: String): StorageObject {
        val resp = transport.requestJson("GET", "$PREFIX/buckets/$bucket/objects/$key", null)
        val b64 = resp.stringField("dataB64")
            ?: throw QnsiApiException("storage.getObject: response missing dataB64", 200)
        val bytes = decodeB64(b64)
            ?: throw QnsiApiException("storage.getObject: dataB64 is not valid base64", 200)
        return StorageObject(bytes, resp)
    }

    public fun deleteObject(bucket: String, key: String) {
        transport.execute("DELETE", "$PREFIX/buckets/$bucket/objects/$key", null, RequestOptions())
    }

    @JvmOverloads
    public fun listObjects(bucket: String, query: Map<String, String?> = emptyMap()): JsonObject =
        transport.requestJson("GET", "$PREFIX/buckets/$bucket/objects", null, RequestOptions(query = query))

    public fun listBuckets(): JsonObject =
        transport.requestJson("GET", "$PREFIX/buckets", null)
}

File: sdks/swift/Sources/QNSI/Clients/StorageClient.swift
import Foundation

private let prefix = "/proxy/storage/v1"

/// Decrypted object payload plus the descriptor JSON returned by the service.
public struct StorageObject: Sendable {
    public let data: Data
    public let descriptor: JSONValue
}

/// PQC-encrypted object storage with SSE-X. Wraps storage-service routes at
/// `/proxy/storage/v1` on the edge gateway.
public struct StorageClient: Sendable {
    let transport: Transport

    public func putObject(
        bucket: String,
        key: String,
        data: Data,
        contentType: String? = nil,
        sseAlgorithm: String? = nil,
        metadata: [String: JSONValue]? = nil,
        idempotencyKey: String? = nil
    ) async throws -> JSONValue {
        var object: [String: JSONValue] = ["dataB64": .string(encodeB64(data))]
        if let contentType { object["contentType"] = .string(contentType) }
        if let sseAlgorithm { object["sseAlgorithm"] = .string(sseAlgorithm) }
        if let metadata { object["metadata"] = .object(metadata) }
        return try await transport.requestJson(
            "PUT", "\(prefix)/buckets/\(bucket)/objects/\(key)", body: .object(object),
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
    }

    /// Fetch and decrypt an object. Throws `QnsiError.api` if the descriptor omits `dataB64`.
    public func getObject(bucket: String, key: String) async throws -> StorageObject {
        let resp = try await transport.requestJson("GET", "\(prefix)/buckets/\(bucket)/objects/\(key)")
        guard let b64 = resp["dataB64"]?.stringValue else {
            throw QnsiError.api(
                message: "storage.getObject: response missing dataB64", statusCode: 200, code: nil, body: nil
            )
        }
        guard let bytes = decodeB64(b64) else {
            throw QnsiError.api(
                message: "storage.getObject: dataB64 is not valid base64", statusCode: 200, code: nil, body: nil
            )
        }
        return StorageObject(data: bytes, descriptor: resp)
    }

    public func deleteObject(bucket: String, key: String) async throws {
        _ = try await transport.execute(
            "DELETE", "\(prefix)/buckets/\(bucket)/objects/\(key)", body: nil, options: RequestOptions()
        )
    }

    public func listObjects(bucket: String, query: [String: String?] = [:]) async throws -> JSONValue {
        try await transport.requestJson(
            "GET", "\(prefix)/buckets/\(bucket)/objects", options: RequestOptions(query: query)
        )
    }

    public func listBuckets() async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/buckets")
    }
}
